parity-crypto/src/secp256k1_alt.rs

```diff
@@ -0,0 +1,411 @@
+
+const ONE_BYTES: [u8;32] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1];
+
+const ZERO_BYTES: [u8;32] = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+
+/// The X coordinate of the generator (could get from lib AFFINE_G const but it is more convenient
+/// this way : could be default value of a trait)
+pub const GENERATOR_X: [u8; 32] = [
+	0x79, 0xbe, 0x66, 0x7e, 0xf9, 0xdc, 0xbb, 0xac,
+	0x55, 0xa0, 0x62, 0x95, 0xce, 0x87, 0x0b, 0x07,
+	0x02, 0x9b, 0xfc, 0xdb, 0x2d, 0xce, 0x28, 0xd9,
+	0x59, 0xf2, 0x81, 0x5b, 0x16, 0xf8, 0x17, 0x98
+];
+
+/// The Y coordinate of the generator
+pub const GENERATOR_Y: [u8; 32] = [
+	0x48, 0x3a, 0xda, 0x77, 0x26, 0xa3, 0xc4, 0x65,
+	0x5d, 0xa4, 0xfb, 0xfc, 0x0e, 0x11, 0x08, 0xa8,
+	0xfd, 0x17, 0xb4, 0x48, 0xa6, 0x85, 0x54, 0x19,
+	0x9c, 0x47, 0xd0, 0x8f, 0xfb, 0x10, 0xd4, 0xb8
+];
+
+/// The order of the secp256k1 curve
+pub const CURVE_ORDER: [u8; 32] = [
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+	0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b,
+	0xbf, 0xd2, 0x5e, 0x8c, 0xd0, 0x36, 0x41, 0x41
+];
+
+lazy_static! {
+	static ref MINUS_ONE_KEY: SecretKey = SecretKey::new(SecretKeyInner::parse(&MINUS_ONE_BYTES).expect("static; qed"));
+	static ref ONE_KEY: SecretKey = SecretKey::new(SecretKeyInner::parse(&ONE_BYTES).expect("static; qed"));
+	static ref ZERO_KEY: SecretKey = SecretKey::new(SecretKeyInner::parse(&ZERO_BYTES).expect("static; qed"));
+}
+
+pub fn one_key() -> &'static SecretKey {
+	&ONE_KEY
+}
+
+pub fn minus_one_key() -> &'static SecretKey {
+	&MINUS_ONE_KEY
+}
+
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct PublicKey(PublicKeyInner);
+
+impl PublicKey {
+
+	fn new(inner: PublicKeyInner) -> Self {
+		PublicKey(inner)
+	}
+
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct SecretKey(SecretKeyInner);
+
+impl Drop for SecretKey {
+	fn drop(&mut self) {
+		// TODO find a way to clear secret, next lines break on mem replace
+		//let key = std::mem::replace(&mut self.0, ZERO_KEY.0.clone());
+		//let buf = &mut Into::<Scalar>::into(*key.inner).0;
+		//Clear::clear(buf);
+	}
+}
+
+impl SecretKey {
+
+	fn new(inner: SecretKeyInner) -> Self {
+		SecretKey(inner)
+	}
+
+}
+
+impl Asym for Secp256k1 {
+	type PublicKey = PublicKey;
+	type SecretKey = SecretKey;
+
+	/// This is highly ethereum opignated
+	const SIGN_SIZE: usize = SIGN_SIZE;
+
+	/// Warning we use 64 bit pubsize (first bytes of 65 bit representation is 4).
+	const PUB_SIZE: usize = PUB_SIZE;
+
+	const SECRET_SIZE: usize = SECRET_SIZE;
+
+	const KEYPAIR_INPUT_SIZE: usize = Self::SECRET_SIZE;
+
+	fn recover(signature: &[u8], message: &[u8]) -> Result<Self::PublicKey, Error> {
+		let mut buf = [0;32];
+		buf.copy_from_slice(&message[..]); // panic on incorrect message size
```

Comment:
Ok I am seeing those commits on your crate, I will use them, it is probably best if I wait a bit, thanks :+1: 

I also got a thing I couldn't do : 
https://github.com/cheme/parity-common/blob/30d91988378d0f0c4c070c01620107e65334950a/parity-crypto/src/secp256k1_alt.rs#L123-L130
Basically if there was an 'unsafe' accessor to the secret key 'Field' I could use Clear on it.

Reply:
Will adding `Default` trait to `SecretKey` work? It looks like in this way it automatically implements `cod::clear::Clear`: https://github.com/sorpaas/libsecp256k1-rs/blob/master/tests/verify.rs#L53

I currently set the "default" secret key to `0, 0, ..., 1` because all zero is not valid.

Reply:
It works (I wish I could do it in this crate to avoid it but my tests where unsuccessfull).